### PR TITLE
Add ClusterInterceptor CRD for registering interceptors

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

This commit adds a new CRD type called ClusterInterceptor. In TEP-0026 this
type was called InterceptorConfig 

The `spec` currently only contains a clientConfig field to locate where
the interceptoris running. Other fields will be added as they are
implemented in follow ups.

This commit also adds a simple reconciler for this type that resolves
the clientConfig to a URL and adds it to the `status.address.url` field.

Part of #868

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
A new CRD called ClusterInterceptor has been added that allows for users to register new Interceptor types.
```

### CODE REVIEW

**Comment by dibyom** [comment]
/cc @savitaashture @wlynch @MarcelMue @bigkevmcd  

I broke off https://github.com/tektoncd/triggers/pull/921 into multiple PRs. One change I did was to call the CRD InterceptorType instead of just Interceptor to indicate that it defines a new type or kind of interceptor.

File: docs/interceptortypes.md
Comment (wlynch) [inline docs/interceptortypes.md:9]: I'm not 100% sure we want this to be cluster scoped. In particular, I'm thinking about how we can expose custom interceptors that can only be used within a single tenant namespace.

File: pkg/apis/triggers/v1alpha1/interceptor_type_types.go
Comment (wlynch) [inline pkg/apis/triggers/v1alpha1/interceptor_type_types.go:35]: Should we reuse [WebhookClientConfig](https://github.com/kubernetes/apiextensions-apiserver/blob/b343cbeb89f1b1f00514360d1e67397207b2e7be/pkg/apis/apiextensions/types.go#L107)? 

At the very least we can likely reuse the ServiceReference if we want to use the apis.URL for URL. The cert data is also useful for HTTPS.

File: docs/interceptortypes.md
Comment (wlynch) [inline docs/interceptortypes.md:23]: bikeshed: I kind of like InterceptorConfiguration to be consistent with [ValidatingWebhookConfiguration](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).

Other good ideas to borrow from admission webhooks we should look into later:
- Version info (e.g. have different routes based on interceptor version)
- Timeout

File: docs/interceptortypes.md
Comment (wlynch) [inline docs/interceptortypes.md:51]: This should be nested in `service`. We should also make note of the optional fields.

File: pkg/apis/triggers/v1alpha1/interceptor_type_defaults.go
Comment (wlynch) [inline pkg/apis/triggers/v1alpha1/interceptor_type_defaults.go:28]: 🤷  0 isn't a valid port, so it probably doesn't matter.

File: pkg/apis/triggers/v1alpha1/interceptor_type_types.go
Comment (MarcelMue) [reply pkg/apis/triggers/v1alpha1/interceptor_type_types.go:35]: I think using a WebhookClientConfig is a good idea - maybe just reimplementing the type here if we are concerned about dependencies.

Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/interceptor_type_types.go:35]: yeah, I am basically reimplementing the type here....two differences:
1. using apis.URL instead of string for the URL type for better validation/json marshalling
2. we do not have a `caCert` field type (yet!) for HTTPS


Pulling in a new dependency for just the two structs  seems much so I decided to copy the structs. It also allows us to make changes (e.g. making the port a *int instead of int as suggested below)

File: docs/interceptortypes.md
Comment (dibyom) [reply docs/interceptortypes.md:51]: 🤦

File: pkg/apis/triggers/v1alpha1/interceptor_type_defaults.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/interceptor_type_defaults.go:28]: I'll switch to a pointer type to make this clearer.

File: docs/interceptortypes.md
Comment (dibyom) [reply docs/interceptortypes.md:9]: If its not cluster scoped, each new namespace would have to install their own version of Interceptors....could we not use RBAC to restrict if certain interceptors need to be limited?

File: docs/interceptortypes.md
Comment (dibyom) [reply docs/interceptortypes.md:23]: hmmm yeah...this CRD is definitely based on the WebhookConfiguration CRDs....I don't have a strong preference on calling this InterceptorConfguration --> I thought it was a bit confusing in the sense its not clear if it is some interceptor configuration that can be reused within a trigger vs configuration for a new kind of interceptor.

File: docs/interceptortypes.md
Comment (dibyom) [reply docs/interceptortypes.md:9]: chatted with @wlynch about this. We will start off with a cluster scoped CRD but will implement a Namespace scoped version of this in the future as well.

File: docs/interceptortypes.md
Comment (wlynch) [reply docs/interceptortypes.md:51]: This still needs to be fixed 😅

File: docs/interceptortypes.md
Comment (wlynch) [reply docs/interceptortypes.md:9]: Thanks! Let's update the docs to reflect the change.

**Review by wlynch — COMMENTED** [review]
/lgtm Looks good!

Approval pending doc changes / any other feedback from others.

File: docs/interceptortypes.md
Comment (MarcelMue) [reply docs/interceptortypes.md:23]: I personally don't love the `Configuration` suffix. It makes sense if we have a different `Interceptor` object but a `Configuration` suffix without an `Interceptor` CRD seems weird to me.

No strong opinions though.

**Review by MarcelMue — COMMENTED** [review]
Docs are out of date, other than that LGTM.

**Comment by dibyom** [comment]
Ok, I think I'm going to update the names to ClusterInterceptor (and Interceptor for the namespaced variant in the future).

File: docs/clusterinterceptors.md
Comment (wlynch) [inline docs/clusterinterceptors.md:35]: Few more Interceptor -> ClusterInterceptor replacements need to be made in here.

**Review by wlynch — COMMENTED** [review]
/lgtm w/ minor doc changes.

Since this is a new type, would appreciate a second pair of eyes as an approver!

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults.go:23]: Just a thought that 80 and 443 are kind of standard port 
So can we use some other port?

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults.go:2]:
```suggestion
Copyright 2021 The Tekton Authors
```

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults_test.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults_test.go:50]: we have only one case so shall we write without using for loop ? i mean direct calling function

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults_test.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults_test.go:1]: missing License

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go:1]: License

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go:59]:
```suggestion
// ServiceReference is a reference to a Service object
```

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go:70]: can we add `omitempty` as its optional field

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go:97]: can we use `defaultPort` from default.go file so that we can avoid hard coding here

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types_test.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_types_test.go:1]: shall we include license here ?

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_validation.go
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/cluster_interceptor_validation.go:2]:
```suggestion
Copyright 2021 The Tekton Authors
```

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go:70]: done

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/cluster_interceptor_types.go:97]: done!

File: pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults.go
Comment (dibyom) [reply pkg/apis/triggers/v1alpha1/cluster_interceptor_defaults.go:23]: users can always specify a different port. I think 80 makes a good default since otherwise every interceptor author will have to specify the "custom" default port that we added. 

(When we add TLS support, we could try 443 as default perhaps -- that's what the WebhookConfiguration CRDs do.)

**Review by savitaashture — COMMENTED** [review]
/lgtm

**Review by savitaashture — APPROVED** [review]

**Pull request merged**
